modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorage.java

```diff
@@ -52,39 +71,134 @@ public abstract class AbstractPageMemoryMvPartitionStorage implements MvPartitio
      * @param versionChainTree Table tree for {@link VersionChain}.
+     * @param indexMetaTree Tree that contains SQL indexes.
      */
     protected AbstractPageMemoryMvPartitionStorage(
             int partitionId,
-            TableView tableView,
-            PageMemory pageMemory,
+            AbstractPageMemoryTableStorage tableStorage,
             RowVersionFreeList rowVersionFreeList,
-            VersionChainTree versionChainTree
+            IndexColumnsFreeList indexFreeList,
+            VersionChainTree versionChainTree,
+            IndexMetaTree indexMetaTree
     ) {
         this.partitionId = partitionId;
+        this.tableStorage = tableStorage;
 
         this.rowVersionFreeList = rowVersionFreeList;
+        this.indexFreeList = indexFreeList;
+
         this.versionChainTree = versionChainTree;
+        this.indexMetaTree = indexMetaTree;
+
+        PageMemory pageMemory = tableStorage.dataRegion().pageMemory();
 
-        groupId = tableView.tableId();
+        int groupId = tableStorage.configuration().value().tableId();
 
         rowVersionDataPageReader = new DataPageReader(pageMemory, groupId, IoStatisticsHolderNoOp.INSTANCE);
     }
 
+    /**
+     * Starts a partition by initializing its internal structures.
+     */
+    public void start() {
+        try {
+            IgniteCursor<IndexMeta> cursor = indexMetaTree.find(null, null);
+
+            NamedListView<? extends TableIndexView> indicesCfgView = tableStorage.configuration().value().indices();
+
+            while (cursor.next()) {
+                IndexMeta indexMeta = cursor.get();
+
+                TableIndexView indexCfgView = getByInternalId(indicesCfgView, indexMeta.id());
+
+                if (indexCfgView instanceof HashIndexView) {
+                    createHashIndex(indexMeta);
+                } else if (indexCfgView instanceof SortedIndexView) {
+                    throw new UnsupportedOperationException("Not implemented yet");
+                } else {
+                    assert indexCfgView == null;
+
+                    //TODO IGNITE-17626 Drop the index synchronously.
+                }
+            }
+        } catch (IgniteInternalCheckedException e) {
+            throw new StorageException("", e);
+        }
+    }
+
+    /**
+     * Returns a hash index instance, creating index it if necessary.
+     *
+     * @param indexId Index UUID.
+     */
+    public HashIndexStorage getOrCreateHashIndex(UUID indexId) {
+        return indexes.computeIfAbsent(indexId, uuid -> createHashIndex(new IndexMeta(indexId, 0L)));
+    }
+
+    private PageMemoryHashIndexStorage createHashIndex(IndexMeta indexMeta) {
+        TableView tableView = tableStorage.configuration().value();
+
+        var indexDescriptor = new HashIndexDescriptor(indexMeta.id(), tableView);
+
+        try {
+            PageMemory pageMemory = tableStorage.dataRegion().pageMemory();
+
+            int groupId = tableView.tableId();
```

Comment:
Why don't we store groupId in a field anymore?

Reply:
I don't know why I did that